### PR TITLE
enhancement(kubernetes_logs source): support for kubeconfigs as an alternative to in_cluster configuration

### PR DESCRIPTION
**Pull request by johanot**
This is my first PR to Vector. First of all, thank you for a great product.

I'm very excited about the Kubernetes support, however I don't plan to run Vector in-cluster, which, looks like the only thing supported at the moment.

This PR implements kubernetes source configuration via a kubeconfig file, which can be formatted as either yaml or json.

If `kube_config_file` is set in the source configuration, then this will be used, otherwise in_cluster configuration will be applied.

Please let me know whether something like this is usable for you. :)

Basic requirements for the kubeconfig current are:

- The kubeconfig file must have at least one context, and a `current-context` must be defined.
- The active context must refer to both a cluster and a user
- The context-referred cluster must exist in the `clusters`-section and the user must exist in the `auth-infos`-section

Supported auth-methods:

- None
- Client-cert file
- Token

Unsupported kubeconfig features:

- In-lined passwords
- In-lined client certs
- OIDC providers
- Extensions
- Unnamed contexts, clusters and users

( above list is probably not exhaustive )

### CODE REVIEW

File: src/kubernetes/client/config/mod.rs
Comment (MOZGIII) [inline src/kubernetes/client/config/mod.rs:8]: Let's move this mod under the `kubeconfig` mod.

**Review by MOZGIII — DISMISSED** [review]
The code submitted in this PR looks great, thanks!

I was initially unsure how this feature could be used but then I realized it can be used to deploy Vector to the Kubernetes cluster nodes at the host to consume logs while being outside of the cluster itself. It's not really a supported way to use `kubernetes_logs` source, but it might be useful for some people.

If we want to merge this, we need an E2E test for it.

The implementation lacks the support for `KUBECONFIG` env var and multiple files, which I would consider mandatory for proper support for this feature. Without it, we'll inevitably be confusing some users, and I'd rather avoid this and implement proper `KUBECONFIG` env var support as part of the base feature (including multifile support). I'd be happy to help with it if you're interested in working on it.

See https://kubernetes.io/docs/concepts/configuration/organize-cluster-access-kubeconfig/ for more info.

**Comment by johanot** [comment]
@MOZGIII 
Thank you for the feedback!

> It's not really a supported way to use `kubernetes_logs` source, but it might be useful for some people.

It would be very useful for us at least. We deny hostPath-volumes entirely in our cluster and lock down all other host-level access for containers, hence we prefer to run Vector as a node-process with systemd.

> we need an E2E test for it.

Will look into it.

> The implementation lacks the support for KUBECONFIG env var

Right. I thought about that too. I might have just made a subconscious choice against it, because I prefer more local scoped configuration. But I agree with your arguments about less confusion for other users. Would you be fine with:

1. drop the `kube_config_file` option
2. create a new option e.g. `kube_config_context` falling back to "current-context"
3. honor `KUBECONFIG` from the environment with support for multiple contexts (and multiple files)

re 1) we can also keep the `kube_config_file` option and just fallback to the env var?

Regarding support for multiple files: I'm personally fine with scoping this to a separate issue/PR, since it involves a bit more work?

**Comment by MOZGIII** [comment]
> 1. drop the `kube_config_file` option
> 2. create a new option e.g. `kube_config_context` falling back to "current-context"
> 3. honor `KUBECONFIG` from the environment with support for multiple contexts (and multiple files)
> 
> re 1) we can also keep the `kube_config_file` option and just fallback to the env var?

I like this concept!

Regarding the configuration though, I think we need to approach it a bit more explicitly: we should have a `struct Kubeconfig` with all the necessary configuration as part of it - i.e. `context: Option<String>` to optionally explicitly specify the context or use `current-context` if not set; the `kubeconfig_files: Option<Vec<PathBuf>>` to override the of the `KUBECONFIG` env var semnatics, or use the `KUBECONFIG` env var / default path if this var is not set:

From the `kubectl `, the default kubeconfig resolution logic is as follows:
```
 The loading order follows these rules:

  1.  If the --kubeconfig flag is set, then only that file is loaded. The flag may only be set once and no merging takes
place.
  2.  If $KUBECONFIG environment variable is set, then it is used as a list of paths (normal path delimiting rules for
your system). These paths are merged. When a value is modified, it is modified in the file that defines the stanza. When
a value is created, it is created in the first file that exists. If no files in the chain exist, then it creates the
last file in the list.
  3.  Otherwise, ${HOME}/.kube/config is used and no merging takes place.
```

We should also have an option to allow configuring whether in-cluster or kubeconfig cluster config is used.

So, in the source config, we'll end up with something like this:

```rust
... Config {
    // ...
    cluster_config: ClusterConfig,
    kubeconfig: KubeconfigParams,
}

enum ClusterConfig {
    InCluster,
    Kubeconfig,
}

struct KubeconfigParams {
    // fields as outlined above
}
```

We'll default to `InCluster` mode, and make `Kubeconfig` an explicit opt-in without the fallbacks to it, or any other kind of autoconfiguration that would pick it implicitly, to make it simple for the end-users. I went through possible autoconfiguration/fallback mechanisms, and it seems like none is obvious or can be considered a default, so for now we'll not going to include one. By meking the `ClusterConfig` an enum, we leave the possibility of introducing an `Auto` mode later.

**Comment by MOZGIII** [comment]
@johanot could you please add new commits rather than amend & force-push - focusing on changes will be easier for me in the review process. Thanks!

**Comment by MOZGIII** [comment]
I've created an issue that covers this use case: https://github.com/timberio/vector/issues/5255. It explores some of the alternatives and might be a useful read.

Looking forward to hearing what you think @johanot!

**Comment by johanot** [comment]
@MOZGIII Great! Thanks for opening the issue. I've been very busy the last couple of weeks, but I hope to get time to look at it soon.

**Comment by binarylogic** [comment]
@johanot thanks for your contribution. What's the status of this PR? Is it blocked and should we try to get this over the finish line?

cc @MOZGIII @eeyun for input.

**Comment by johanot** [comment]
> Is it blocked and should we try to get this over the finish line?

@binarylogic I'd love to. Unfortunately, I've been very busy lately and haven't got the time for the rewrite proposed previously in this thread. If someone else can find the time, I'd be happy to review or test.

**Review by eeyun — APPROVED** [review]
So here's where I'm at on this. This PR does what it set out to do and it solves an immediate problem. It doesn't completely cover the edge case of multiple files as @MOZGIII mentioned but I think this can and should be merged as is. We're already tracking issues for extending the e2e test-harness to allow for node-level testing and we can add a new issue to track adding implementation to support the multi-file edge case.

**Comment by eeyun** [comment]
@johanot could you rebase this against master or click the box that allows us to push to your branch so we can rebase it and get it merged?

**Comment by johanot** [comment]
@eeyun "allow edits by maintainers" is already checked, thanks :)

**Review by lukesteensen — COMMENTED** [review]
I don't have as much k8s context, but the code here looks great! We do need to check out the licensing situation before merging, but otherwise 👍

File: src/kubernetes/client/config/kubeconfig_types.rs
Comment (johanot) [reply src/kubernetes/client/config/kubeconfig_types.rs:3]: Just wanted to add the full disclosure with a code-comment, that this file is a 1:1 copy-paste. I have to fully admit that I didn't check the licensing details, and of course you're right. We need to ensure compatibility here.

Comment (eeyun) [reply src/kubernetes/client/config/kubeconfig_types.rs:3]: Kube-rs is apache licensed.

Comment (eeyun) [reply src/kubernetes/client/config/kubeconfig_types.rs:3]: @johanot would you like me to implement these cleanup items? I'm happy to finish carrying this across the finish line if you need the support, just let me know :)

Comment (johanot) [reply src/kubernetes/client/config/kubeconfig_types.rs:3]: @eeyun that would be great help, thanks! :)

File: src/kubernetes/client/config/kubeconfig.rs
Comment (eeyun) [reply src/kubernetes/client/config/kubeconfig.rs:66]: It's very possible that I'm missing something, but ultimately resolving the type errors when converting this to combinators leaves me doing battle with the borrow checker. If you're not vehemently opposed to the styling of nested matches I'd personally opt to leave this as is.

Comment (eeyun) [reply src/kubernetes/client/config/kubeconfig.rs:66]: The specific issue is because of the calling `read_to_string` inside of a map callback where we need to handle the result. We can implicitly convert it to an option but doing so leaves us with an `Option<Option<String>>` where our config struct expects an `Option<String>` then of course just unwrapping that is a no go because the user sits behind a shared ref.

**Comment by eeyun** [comment]
Apologies. This needed one last rebase before merge. Once these tests pass I _believe_ this should be good to go.

**Comment by johanot** [comment]
Thank you for all your work @eeyun ! :) It looks like this is ready for merge now?

**Comment by eeyun** [comment]
You bet. Thanks for your contribution!

**Pull request merged**
